src/gfx/pal_packing.cpp

```diff
@@ -0,0 +1,359 @@
+		using value_type = ProtoPalAttrs;
+		using pointer = Constness<value_type> *;
+		using reference = Constness<value_type> &;
+		using iterator_category = std::input_iterator_tag;
+
+	private:
+		Constness<decltype(_assigned)> *_array = nullptr;
+		Inner _iter{};
+
+		Iter(decltype(_array) array, decltype(_iter) &&iter) : _array(array), _iter(iter) {
+			skipEmpty();
+		}
+		void skipEmpty() {
+			while (_iter != _array->end() && !(*_iter).has_value()) {
+				++_iter;
+			}
+		}
+
+	public:
+		Iter() = default;
+
+		bool operator==(Iter const &other) const { return _iter == other._iter; }
+		bool operator!=(Iter const &other) const { return !(*this == other); }
+		Iter &operator++() {
+			++_iter;
+			skipEmpty();
+			return *this;
+		}
+		Iter operator++(int) {
+			Iter it = *this;
+			++(*this);
+			return it;
+		}
+		reference operator*() const {
+			assert((*_iter).has_value());
+			return **_iter;
+		}
+		pointer operator->() const {
+			return &(**this); // Invokes the operator above, not quite a no-op!
+		}
+
+		friend void swap(Iter &lhs, Iter &rhs) {
+			swap(lhs._array, rhs._array);
+			swap(lhs._iter, rhs._iter);
+		}
+	};
+public:
+	using iterator = Iter<decltype(_assigned)::iterator, std::remove_const_t>;
+	iterator begin() { return iterator{&_assigned, _assigned.begin()}; }
+	iterator end() { return iterator{&_assigned, _assigned.end()}; }
+	using const_iterator = Iter<decltype(_assigned)::const_iterator, std::add_const_t>;
+	const_iterator begin() const { return const_iterator{&_assigned, _assigned.begin()}; }
+	const_iterator end() const { return const_iterator{&_assigned, _assigned.end()}; }
+
+	/**
+	 * Assigns a new ProtoPalAttrs in a free slot, assuming there is one
+	 * Args are passed to the `ProtoPalAttrs`'s constructor
+	 */
+	template<typename... Ts>
+	auto assign(Ts &&...args) {
+		auto freeSlot = std::find_if_not(
+			_assigned.begin(), _assigned.end(),
+			[](std::optional<ProtoPalAttrs> const &slot) { return slot.has_value(); });
+
+		if (freeSlot == _assigned.end()) { // We are full, use a new slot
+			_assigned.emplace_back(std::forward<Ts>(args)...);
+		} else { // Reuse a free slot
+			(*freeSlot).emplace(std::forward<Ts>(args)...);
+		}
+		return freeSlot;
+	}
+	void remove(iterator const &iter) {
+		(*iter._iter).reset(); // This time, we want to access the `optional` itself
+	}
+	void clear() { _assigned.clear(); }
+
+	/**
+	 * Computes the "relative size" of a proto-palette on this palette
+	 */
+	double relSizeOf(ProtoPalette const &protoPal) const {
+		return std::transform_reduce(
+			protoPal.begin(), protoPal.end(), .0, std::plus<>(), [this](uint16_t color) {
+				// NOTE: The paper and the associated code disagree on this: the code has
+			    // this `1 +`, whereas the paper does not; its lack causes a division by 0
+			    // if the symbol is not found anywhere, so I'm assuming the paper is wrong.
+				return 1.
+			           / (1
+			              + std::count_if(
```

Comment:
This isn't clang-formatted, is it? No need to do so yet, but I'd be concerned if that's what our auto-formatter does. :P

Reply:
It is.

Reply:
Here is the problem: (dots are spaces)

![image](https://user-images.githubusercontent.com/15271137/156226067-c8a806c5-211d-499e-9d7f-c40c86c2184d.png)

Reply:
Hmm. I thought it would normalize all leading whitespace to our expected indent-with-tabs, align-with-spaces style. Is it preserving spaces we just happened to have in those lines after tabs?

Reply:
Nope, I changed the spaces to tabs, reformatted, and the spaces came back. This might have something to do with our `.clang-format`, I'll investigate further.

Reply:
I think we should use `UseTab: ForIndentation`

Reply:
`ForContinuationAndIndentation` also [sounds plausible](https://clang.llvm.org/docs/ClangFormatStyleOptions.html). Let's see what each does with this code.

Reply:
What I hope `clang-format` will do is, for the multi-line contents of things like that `std::count_if(...)` call, only indent the contents by a single tab, don't try and align them to the open parenthesis. That should keep deeply-nested code like this from getting ridiculous.

Reply:
That's `AlignAfterOpenBracket`, though.

Reply:
I mean, part of that screenshot looks like what I'd expect from the name "`AlignAfterOpenBracket`", but that's not what we're using...

Reply:
Also, this only appears ridiculous because my editor currently has tabs set to 8 wide (for the C code still using `checkpatch` style); with 4, it's much better.

![Screenshot](https://user-images.githubusercontent.com/15271137/156812694-56bb79c6-9312-4b96-bbf6-9ed9616322d2.png)

Reply:
> I mean, part of that screenshot looks like what I'd expect from the name "`AlignAfterOpenBracket`", but that's not what we're using...

Er, but we are?

https://github.com/ISSOtm/rgbds/blob/75d54f439492a960ae5930aa26797789cd244ffe/.clang-format#L2

Reply:
Whatever value `UseTab` has should be chosen so tab width doesn't matter though.

Reply:
(If nothing gets this to look decent, maybe do `auto n = std::count_if(...); return 1.0 / (1 + n);`.)

Reply:
> Er, but we are?

Oops, I thought you mean that `AlignAfterOpenBracket` was our value of `UseTab`, for which we have `UseTab: AlignWithSpaces`.

If the C++ is regularly going to have deeply-nested brackets like this, I'd like to either avoid that with more local variables, or not use `AlignAfterOpenBracket`.

Reply:
I'll advocate for the latter if only because of lambdas (check the weird alignment out)

Reply:
Do `UseTab: ForIndentation` or `UseTab: ForContinuationAndIndentation` give better-looking results to you?